### PR TITLE
Support `git show **`

### PR DESCRIPTION
**Pull request by hexium310**
## Fixed
- When it was selected from the completions of `git push origin master:ma**`, `master:mamaster` was inserted

## Features
```sh
$ git show **
$ git show [<rev>...] -- **
$ git show <rev>:**
```
This PR makes it returns all arguments list when `0` is passed to 2nd argument of `_fzf_complete_parse_argument`.

https://git-scm.com/docs/git-show

### CODE REVIEW

File: tests/git.zunit [hunk 1]
```diff
@@ -4299,6 +4299,212 @@
...
+    _fzf_complete_git 'git show '
+}
+
+@test 'Testing completion: git show another-branch:file1 master new-branch v1:file1 -- **' {
+    run git checkout -b new-branch
+    echo >> newfile1
+    echo >> newfile2
+    run git add .
+    run git commit -m 'new commit'
+    run git checkout master
```
Comment (chitoku-k) [inline tests/git.zunit:4438]: Do not do this.

File: tests/git.zunit [hunk 2]
```diff
@@ -4299,6 +4299,212 @@
...
+
+    _fzf_complete() {
+        assert $# equals 6
+        assert $1 same_as '--ansi'
+        assert $2 same_as '--read0'
+        assert $3 same_as '--print0'
+        assert $4 same_as '--multi'
+        assert $5 same_as Make it returns all arguments list when `0` is passed to 2nd argument of `_fzf_complete_parse_argument`.'--'
```
Comment (hexium310) [inline tests/git.zunit:4446]: It is the mistake🙇‍♂️

**Comment by chitoku-k** [comment]
This patch generally looks OK. Nice work👌

I found an edge case where an empty item is listed if multiple treeish objects are specified before `--` like the following:

```bash
[lime  fzf-zsh-completions | test  !  $] git show 8b7835d e0ca3c3 7b1302c -- **
>
  6/6 (0)
>
  src/completers/git.zsh
  src/completers/kubectl.zsh
  src/core/parser.zsh
  tests/git.zunit
  tests/kubectl.zunit
```

This is caused by the behaviour in `git show` that prints an empty splitter for each object it receives (`^@` indicates NUL):
```bash
[lime  fzf-zsh-completions | test  !  $] git show --pretty=format: --name-only -z 8b7835d e0ca3c3 7b1302c
src/completers/kubectl.zsh^@src/core/parser.zsh^@^@src/completers/kubectl.zsh^@tests/kubectl.zunit^@^@src/completers/git.zsh^@tests/git.zunit^@
                                               ^^^^                                               ^^^^
                                               Note here empty splitters are seen.
```

File: src/completers/git.zsh [hunk 3]
```diff
@@ -511,6 +511,114 @@ _fzf_complete_git() {
...
+            --encoding --expand-tabs --find-object --format --inter-hunk-context
+            --line-prefix --output --output-indicator-context --output-indicator-new
+            --output-indicator-old --src-prefix --unified --word-diff-regex --ws-error-highlight
+        )
+        local git_options_argument_optional=(
+            -B -C -M -X
+            --abbrev --break-rewrites --color --color-moved --dirstat
+            --find-copies --find-renames --ignore-submodules --notes
+            --pretty --relative --show-notes --stat --submodule --word-diff
+        )
```
Comment (chitoku-k) [inline src/completers/git.zsh:528]: No need to break lines. Either write in a single line or write every item in different lines.

File: src/completers/git.zsh [hunk 4]
```diff
@@ -511,6 +511,114 @@ _fzf_complete_git() {
...
+            prefix=${prefix#$prefix_option}
+        fi
+
+        local prefix_ref=${prefix%%[^:]#}
+
+        case $completing_option in
+            --abbrev|--anchored|--break-rewrites|--dst-prefix|--encoding|--expand-tabs| \
+                --find-copies|--find-object|--find-renames|--format|--inter-hunk-context| \
+                --line-prefix|--output|--output-indicator-context|--output-indicator-new| \
+                --output-indicator-old|--src-prefix|--stat|--submodule|--unified|--word-diff-regex)
```
Comment (chitoku-k) [inline src/completers/git.zsh:545]: Here as well.

File: src/core/parser.zsh [hunk 5]
```diff
@@ -95,6 +95,10 @@ _fzf_complete_parse_argument() {
         command_arguments+=${arguments[$i]}
     done
 
+    if [[ $index = 0 ]]; then
+        echo - ${command_arguments}
+        return 0
```
Comment (chitoku-k) [inline src/core/parser.zsh:100]:
```suggestion
        return
```

File: tests/git.zunit [hunk 6]
```diff
@@ -4299,6 +4299,213 @@
...
+        assert ${actual1[3]} same_as 'directory2/file4'
+
+        actual2=(${(0)lines[2]})
+        assert ${#actual2} equals 1
+        assert ${actual2[1]} same_as 'containing'
+
+        actual3=(${(0)lines[3]})
+        assert ${#actual3} equals 2
+        assert ${actual3[1]} same_as 'newlines'
+        assert ${actual3[2]} same_as 'file1'
```
Comment (chitoku-k) [inline tests/git.zunit:4458]: Probably this case should cover the aforementioned bug though it didn't fail...

File: src/completers/git.zsh [hunk 7]
```diff
@@ -689,6 +828,38 @@ _fzf_complete_git-refs_post() {
...
+_fzf_complete_git-notes_post() {
+    awk '{ print $1 }'
+}
+
+_fzf_complete_git-show-files() {
+    local fzf_options=$1
+    shift
+
+    _fzf_complete --ansi --read0 --print0 ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option < <(
+        git show --pretty=tformat: --name-only -z ${(ps: :)treeish} 2> /dev/null | sort -u -z
```
Comment (chitoku-k) [inline src/completers/git.zsh:850]: How about using built-in functionarity in Zsh for deduplication? Probably it should make this faster and more platform-compatible.

File: tests/git.zunit [hunk 8]
```diff
@@ -4299,6 +4299,213 @@
...
+        assert $1 same_as '--ansi'
+        assert $2 same_as '--tiebreak=index'
+        assert $3 same_as '--'
+        assert $4 same_as 'git show --show-notes='
+
+        run cat
+        assert ${#lines} equals 2
+
+        assert ${lines[1]} same_as "${fg[yellow]}notes/commits$reset_color  Notes added by 'git notes add'"
+        assert ${lines[2]} same_as "${fg[yellow]}notes/note   $reset_color  Notes added by 'git notes add'"
```
Comment (chitoku-k) [inline tests/git.zunit:4480]: From out of curiousity, is it technically possible to show the content of notes in completion such as `Note1`?

File: src/completers/git.zsh (hunk 7 above)
Comment (chitoku-k) [reply src/completers/git.zsh:850]: Working example looks like:
```zsh
local result=$(git show --pretty=format: --name-only -z ${(ps: :)treeish} 2> /dev/null)
echo -n ${(pj:\0:)${(u)${(0)result}}}
```

File: tests/git.zunit (hunk 8 above)
Comment (hexium310) [reply tests/git.zunit:4480]: One notes ref is related with some commits. It is maybe difficult to display multiple notes message such as `Note1`.

Comment (chitoku-k) [reply tests/git.zunit:4480]: OK.

**Review by chitoku-k — APPROVED** [review]
Talented!
